### PR TITLE
[mbedtls] free mbedtls on initialization failure

### PR DESCRIPTION
**Pull request by bukepo**
`Dtls::Start` might fails for various reasons, data allocated when trying
initializing mbedtls are not freed. This PR addresses this issue.

This PR also moves initialization of pk and x509 certificate which is freed in `Dtls::Close()` from constructor to `Dtls::Start()`.

### CODE REVIEW

**Comment by bukepo** [comment]
Thanks to @hjian2017 for reporting this issue.

File: src/core/meshcop/dtls.cpp [hunk 1]
```diff
@@ -317,23 +361,7 @@ void Dtls::Close(void)
     mTimer.Start(kGuardTimeNewConnectionMilli);
     VerifyOrExit(mStarted);
```
Comment (hjian2017) [inline src/core/meshcop/dtls.cpp]: If Dtls::Start fails before mStarted = true , please check you will release the resource or not.

**Comment by jwhui** [comment]
It should be safe to call `mbedtls_*_free()` on memory that has been zeroed out. Would it be simpler to ensure that the memory is zeroed out on init and simply call `mbedtls_*_free()` on everything when shutting down DTLS, rather than maintaining state of what has been initialized and what has not?

**Comment by bukepo** [comment]
I am not sure about that, I think it depends on the implementation. Is it safer to rely on documented behaviors?

**Comment by jwhui** [comment]
I came to the current thinking bsed on the [mbedTLS coding standards](https://tls.mbed.org/kb/development/mbedtls-coding-standards):

> The free() function must free any allocated memory within the context, but not the context itself; it must zeroize any data in the context or substructures.

Which indicates any pointer within a given context should be `0`. Calling `mbedtls_free()` on any NULL pointer should have no effect.

Thoughts?

**Comment by bukepo** [comment]
Thanks for the clarification. @jwhui I have updated this PR.

**Pull request merged**